proposals/p2188.md

```diff
@@ -0,0 +1,456 @@
+    -   [Generic bindings](#generic-bindings)
+    -   [`auto`](#auto)
+    -   [`var`](#var)
+    -   [Tuple patterns](#tuple-patterns)
+    -   [Struct patterns](#struct-patterns)
+    -   [Choice patterns](#choice-patterns)
+    -   [Templates](#templates)
+-   [Rationale](#rationale)
+-   [Alternatives considered](#alternatives-considered)
+    -   [Struct pattern syntax](#struct-pattern-syntax)
+-   [Future work](#future-work)
+    -   [Or patterns](#or-patterns)
+    -   [Guards](#guards)
+    -   [User-defined pattern matching](#user-defined-pattern-matching)
+    -   [Matching classes with struct patterns](#matching-classes-with-struct-patterns)
+
+<!-- tocstop -->
+
+## Abstract
+
+This paper proposes concrete syntax and semantic choices for Carbon patterns.
+
+## Problem
+
+Carbon uses patterns wherever a value should be given a name, decomposed, or
+matched against: in function parameters, variable declarations, `match`
+statement `case`s, `for` statement loop variables, and so on. Simplified forms
+of patterns, required to be just a simple name binding, appear in additional
+contexts, such as fields in classes and implicit parameter lists. While we have
+syntax specified for some of these constructs, we do not have an approved
+proposal describing the syntax or semantics of patterns.
+
+## Background
+
+See [pattern matching](https://en.wikipedia.org/wiki/Pattern_matching) on
+wikipedia for a broad overview of the subject.
+
+We refer to the value being matched by a pattern as the _scrutinee_.
+
+## Proposal
+
+Patterns in Carbon are a generalization of the expression grammar. Compared to
+expressions, patterns add:
+
+-   Bindings, of the form `name: type`, which give a name for the scrutinee.
+-   `var` _pattern_, which requests mutable storage be provided for the
+    scrutinee.
+-   Additional syntax to make matching against structs more convenient.
+
+## Details
+
+### Expressions versus proper patterns
+
+A pattern that contains within it any pattern-specific syntax, such as a
+binding, is a _proper pattern_. Any other pattern must necessarily be an
+expression. Many expression forms, such as arbitrary function calls, are not
+permitted as proper patterns, so cannot contain bindings.
+
+-   _pattern_ ::= _proper-pattern_
+
+```
+fn F(n: i32) -> i32 { return n; }
+
+match (F(42)) {
+  // ❌ Error: binding can't appear in a function call.
+  case (F(n: i32)) => {}
+}
+```
+
+### Expression patterns
+
+An expression is a pattern.
+
+-   _pattern_ ::= _expression_
+
+The pattern is compared with the expression using the `==` operator: _pattern_
+`==` _scrutinee_.
+
+```
+fn F(n: i32) {
+  match (n) {
+    // ✅ Results in an `n == 5` comparison.
+    // OK despite `n` and `5` having different types.
+    case 5 => {}
+  }
+}
+```
+
+### Bindings
+
+A name binding is a pattern.
+
+-   _binding-pattern_ ::= `unused`? _identifier_ `:` _expression_
+-   _binding-pattern_ ::= `unused`? _identifier_ `:` _proper-pattern_
```

Comment:
My main arguments against it are readability, and consistency with the decision we made for `impl` declarations. How does the alternative read?
```
fn F[template T:! Type](x: T) -> auto {
  match (x) {
    case forall [U:! Type, V:! AddableWith(U)] (a: U, b: V) => {
      return F(a + b);
    }
    // ...
  }
}
```

@mconst

Reply:
I think I don't have a strong preference here.  I do find `case (a: (U:! Type), b: (V:! AddableWith(U)))` hard to read, but I'm not sure `case forall [U:! Type, V:! AddableWith(U)] (a: U, b: V)` is significantly better, especially since it requires the unusual `forall` keyword.  So either way's fine with me.